### PR TITLE
feat(aiohttp): add client integration

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
# Integration
This PR adds support for [`aiohttp`](https://docs.aiohttp.org/en/stable/client.html) client.

Work was long ago started on this in (https://github.com/DataDog/dd-trace-py/pull/294 (thank you @thehesiod!!!). There were a number of issues in the library that led to not being able to merge in the work such as async context management, service naming, integration configuration and more which have all since been addressed.

#294 and later #1372 included additional support for `ClientResponse` and `StreamReader` which are omitted here with the intention of introducing them as follow ups. 

## Links

- Integration docs: [aiohttp client](https://docs.aiohttp.org/en/stable/client.html)
- Corp docs PR: https://github.com/DataDog/documentation/pull/13333

## Checklist

- [x] Usage and configuration documentation added in `__init__.py`, `docs/index.rst` and `docs/integrations.rst`.
- [x] [Corp docs](https://github.com/Datadog/documentation) PR to add new integration to documentation.
- [x] Span metadata
  - [x] Service (use [`int_service`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L55) or [`ext_service`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L87)).
  - [x] Span type should be one of [these](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/ext/__init__.py#L7).
  - [x] Resource
  - [x] Measured tag
  - [ ] Sample analytics
- [x] Global configuration
  - [x] `ddtrace.config` entry is specified.
  - [x] Environment variables are provided for config options.
- [x] Instance configuration
  - [x] Pin overriding.
  - [x] Service name override (if applicable).
- [x] Async
  - [x] Span parenting behaves as expected.
  - [x] Context propagation across async contexts.
- [x] HTTP
  - [x] Distributed tracing propagation is implemented.
  - [x] Use [`trace_utils.set_http_meta`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L143-L152) to set http tags
- [x] Tests
  - [x] Use `pytest` fixtures found in `tests/conftest.py` or the test helpers on `TracerTestCase` if
    writing `unittest` style test cases.
  - [x] Tests are provided for all the above.
  - [x] Tests are added to CI (`.circleci/config.yml`).
  - [x] Functionality is maintained from original library.
  - [x] Patch test cases are added (see `test_django_patch.py` for an example).
  - [x] All Python versions that the library supports are tested.
  - [x] All significant library versions (including the latest) are tested. This typically includes every minor release going back a few years.

### CODE REVIEW

File: ddtrace/contrib/aiohttp/__init__.py [hunk 1]
```diff
@@ -1,7 +1,36 @@
...
+Or use :func:`patch()<ddtrace.patch>` to manually enable the integration::
+
+    from ddtrace import patch
+    patch(aiohttp=True)
+
+
+Global Configuration
+~~~~~~~~~~~~~~~~~~~~
+
+.. py:data:: ddtrace.config.aiohttp['distributed_tracing']
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/__init__.py:25]: do we need "client" specific naming for config?

```python
ddtrace.config.aiohttp_client["distributed_tracing"]
```

?

since we do have a server component as well

File: ddtrace/contrib/aiohttp/patch.py [hunk 2]
```diff
@@ -1,54 +1,127 @@
...
-
-    template_module = True
-except ImportError:
-    template_module = False
-except Exception:
-    log.warning("aiohttp_jinja2 could not be imported and will not be instrumented.", exc_info=True)
-    template_module = False
 
 config._add(
     "aiohttp",
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:27]:
```suggestion
    "aiohttp_client",
```

Comment (brettlangdon) [reply ddtrace/contrib/aiohttp/patch.py:27]: then would need to keep `aiohttp` for server

File: ddtrace/contrib/aiohttp/patch.py [hunk 3]
```diff
@@ -1,54 +1,127 @@
...
+from ..trace_utils import unwrap
+from ..trace_utils import wrap
 
 
 log = get_logger(__name__)
 
-try:
-    # instrument external packages only if they're available
-    import aiohttp_jinja2
+propagator = HTTPPropagator()
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:23]: why create an instance? they are just static methods

File: releasenotes/notes/aiohttp-98ae9ce70dda1dbc.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    aiohttp: rewrite integration.
```
Comment (brettlangdon) [inline releasenotes/notes/aiohttp-98ae9ce70dda1dbc.yaml:4]: add client tracing support?

File: ddtrace/contrib/aiohttp/patch.py [hunk 5]
```diff
@@ -1,54 +1,127 @@
...
+        set_http_meta(
+            span,
+            config.aiohttp,
+            method=method,
+            url=str(url),
+            query=url.query,
+            request_headers=headers,
+        )
+
+        resp = await func(*args, **kwargs)  # ClientResponse
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:75]: do we need to add headers back into `kwargs`? if they aren't in kwargs, then we create a new dict, and then it won't get used ?

File: ddtrace/contrib/aiohttp/patch.py [hunk 6]
```diff
@@ -1,54 +1,127 @@
...
-    Patch aiohttp third party modules:
-        * aiohttp_jinja2
-    """
-    if template_module:
-        if getattr(aiohttp_jinja2, "__datadog_patch", False):
-            return
-        setattr(aiohttp_jinja2, "__datadog_patch", True)
+    import aiohttp
+
+    if getattr(aiohttp, "_datadog_patch", False):
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:111]: are we missing the aiohttp_jinja2 patching?

File: tests/snapshots/tests.contrib.aiohttp.test_aiohttp_client.test_200_request.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+       "service": null,
+       "resource": "TCPConnector.connect",
+       "trace_id": 0,
+       "span_id": 2,
+       "parent_id": 1,
+       "duration": 1728000,
+       "start": 1646300282383136000
+     },
+        {
+          "name": "TCPConnector._create_connection",
```
Comment (brettlangdon) [inline tests/snapshots/tests.contrib.aiohttp.test_aiohttp_client.test_200_request.json:38]: do we need to trace these both if they are nested?

File: tests/contrib/aiohttp/test_aiohttp_client.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            async with session.get(URL_500) as resp:
+                assert resp.status == 500
+
+
+@pytest.mark.asyncio
+async def test_200_request_distributed_tracing(snapshot_context):
+    with snapshot_context():
+        async with aiohttp.ClientSession() as session:
+            pin = Pin.get_from(session)
+            pin._config["distributed_tracing"] = True
```
Comment (brettlangdon) [inline tests/contrib/aiohttp/test_aiohttp_client.py:46]: isn't the default `True` ?

do we need to test with it being `False`?

File: ddtrace/contrib/aiohttp/__init__.py (hunk 1 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/aiohttp/__init__.py:25]: Yeah since the majority of users are probably using the client and the `aiohttp` entry already exists I opted to stick with the name `aiohttp` but I do like being more explicit. Let's opt for `aiohttp_client` 👍

File: ddtrace/contrib/aiohttp/patch.py (hunk 3 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/aiohttp/patch.py:23]: yep, this is old code

File: ddtrace/contrib/aiohttp/patch.py (hunk 5 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/aiohttp/patch.py:75]: `inject` mutates the value passed to it but correct, in the case when headers are not supplied then this is a bug. Good catch!

File: tests/snapshots/tests.contrib.aiohttp.test_aiohttp_client.test_200_request.json (hunk 7 above)
Comment (Kyle-Verhoog) [reply tests/snapshots/tests.contrib.aiohttp.test_aiohttp_client.test_200_request.json]: I actually removed this instrumentation but didn't update the snapshot here. It will be a follow-up task for whether or not we want to instrument the connection layer.

File: ddtrace/contrib/aiohttp/patch.py [hunk 9]
```diff
@@ -1,25 +1,130 @@
...
-log = get_logger(__name__)
 
+log = get_logger(__name__)
 
 config._add(
-    "aiohttp",
+    "aiohttp_client",
     dict(
-        distributed_tracing=True,
+        distributed_tracing_enabled=asbool(os.getenv("DD_AIOHTTP_CLIENT_DISTRIBUTED_TRACING", False)),
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:27]:
```suggestion
        distributed_tracing_enabled=asbool(os.getenv("DD_AIOHTTP_CLIENT_DISTRIBUTED_TRACING", True)),
```

File: ddtrace/contrib/aiohttp/patch.py [hunk 10]
```diff
@@ -1,25 +1,130 @@
...
+    async def _create_connection(self, req, *args, **kwargs):
+        pin = Pin.get_from(self)
+        with pin.tracer.trace("%s._create_connection" % self.__class__.__name__):
+            result = await self.__wrapped__._create_connection(req, *args, **kwargs)
+            return result
+
+
+@with_traced_module
+async def _traced_clientsession_request(aiohttp, pin, func, instance, args, kwargs):
+    method: str = get_argument_value(args, kwargs, 0, "method")
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:52]:
```suggestion
    method = get_argument_value(args, kwargs, 0, "method")  # type: str
```

?

I know this only works on newer Python, but still should we prefer comment type annotations?

File: ddtrace/contrib/aiohttp/patch.py [hunk 11]
```diff
@@ -1,25 +1,130 @@
...
+    url = URL(get_argument_value(args, kwargs, 1, "url"))  # url can be a str or yarl.URL
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:53]: no type hinting here?

File: ddtrace/contrib/aiohttp/patch.py [hunk 12]
```diff
@@ -1,25 +1,130 @@
...
+    headers = kwargs.get("headers") or {}
+
+    # Create a new span and attach to this instance (so we can
+    # retrieve/update/close later on the response)
+    # Note that we aren't tracing redirects
+    with pin.tracer.trace(
+        "aiohttp.request", span_type=SpanTypes.HTTP, service=ext_service(pin, config.aiohttp)
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:60]:
```suggestion
        "aiohttp.request", span_type=SpanTypes.HTTP, service=ext_service(pin, config.aiohttp_client)
```

?

Comment (brettlangdon) [reply ddtrace/contrib/aiohttp/patch.py:60]: or `pin._config` ?

File: ddtrace/contrib/aiohttp/patch.py [hunk 13]
```diff
@@ -1,25 +1,130 @@
...
+        set_http_meta(
+            span,
+            config.aiohttp_client,
+            method=method,
+            url=str(url),
+            query=url.query,
+            request_headers=headers,
+        )
+
+        resp = await func(*args, **kwargs)  # ClientResponse
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:74]:
```suggestion
        resp = await func(*args, **kwargs)  # type: ClientResponse
```

?

File: ddtrace/contrib/aiohttp/patch.py [hunk 14]
```diff
@@ -1,25 +1,130 @@
...
+
+@with_traced_module_sync
+def _traced_clientsession_init(aiohttp, pin, func, instance, args, kwargs):
+    func(*args, **kwargs)
+    instance._connector = _WrappedConnectorClass(instance._connector, pin)
+
+
+def _patch(aiohttp):
+    Pin().onto(aiohttp)
+    pin = Pin(_config=config.aiohttp_client.copy())
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:90]: do we do `.copy()` on others? 🤔 it makes sense, but if you make a global change it won't apply to anything using this pin?

File: releasenotes/notes/aiohttp-98ae9ce70dda1dbc.yaml [hunk 15]
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    Add aiohttp client integration. This integration traces requests made using the aiohttp client and includes support
```
Comment (brettlangdon) [inline releasenotes/notes/aiohttp-98ae9ce70dda1dbc.yaml:4]:
```suggestion
    aiohttp: Add aiohttp client integration. This integration traces requests made using the aiohttp client and includes support
```

?

File: ddtrace/contrib/async_trace_utils.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    Usage::
+
+        @with_traced_module
+        async def my_traced_wrapper(django, pin, func, instance, args, kwargs):
+            # Do tracing stuff
+            pass
+
+        def patch():
+            import django
+            wrap(django.somefunc, my_traced_wrapper(django))
```
Comment (thehesiod) [inline ddtrace/contrib/async_trace_utils.py:19]: why is there django here, should be more relevant

File: ddtrace/contrib/aiohttp/patch.py (hunk 12 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/aiohttp/patch.py:63]: `ext_service` will check the pin

File: ddtrace/contrib/aiohttp/patch.py (hunk 14 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/aiohttp/patch.py:99]: Not sure how consistent we've been. Copy makes sense, else we'd have to have logic to discern whether a value was user-set or not.

> but if you make a global change it won't apply to anything using this pin

This is already the case for most/all of our global configuration (that it can't change at runtime).

File: ddtrace/contrib/aiohttp/patch.py [hunk 17]
```diff
@@ -1,25 +1,139 @@
...
+
+@with_traced_module
+async def _traced_clientsession_request(aiohttp, pin, func, instance, args, kwargs):
+    method = get_argument_value(args, kwargs, 0, "method")  # type: str
+    url = URL(get_argument_value(args, kwargs, 1, "url"))  # type: Union[str, URL]
+    params = kwargs.get("params") or {}
+    headers = kwargs.get("headers") or {}
+
+    # Create a new span and attach to this instance (so we can
+    # retrieve/update/close later on the response)
```
Comment (brettlangdon) [inline ddtrace/contrib/aiohttp/patch.py:60]: Where do we attach it to the instance?

File: tests/contrib/aiohttp/test_aiohttp_client.py [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+@pytest.mark.parametrize("variant", ["pin", "global"])
+async def test_distributed_tracing_disabled(ddtrace_run_python_code_in_subprocess, variant):
+    code = f"""
+import asyncio
+import sys
+import aiohttp
+from ddtrace import Pin
+from tests.contrib.aiohttp.test_aiohttp_client import URL_200
+
+class MockRequest(object):
```
Comment (brettlangdon) [inline tests/contrib/aiohttp/test_aiohttp_client.py:74]: why not just ping `httpbin` and assert the headers sent from that?

Comment (brettlangdon) [reply tests/contrib/aiohttp/test_aiohttp_client.py:74]: I suppose it really only saves you this mock, otherwise the test is basically the same

File: tests/contrib/aiohttp/test_aiohttp_client.py [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+async def test_500_request(snapshot_context):
+    with snapshot_context():
+        async with aiohttp.ClientSession() as session:
+            async with session.get(URL_500) as resp:
+                assert resp.status == 500
+
+
+@pytest.mark.asyncio
+async def test_200_request_distributed_tracing(tracer, test_spans):
+    class MockRequest(object):
```
Comment (brettlangdon) [inline tests/contrib/aiohttp/test_aiohttp_client.py:45]: same here, why not httpbin?

Comment (brettlangdon) [reply tests/contrib/aiohttp/test_aiohttp_client.py:45]: HTTPretty? which we already use elsewhere

**Review by brettlangdon — COMMENTED** [review]
some minor nits, otherwise looking good

File: ddtrace/contrib/aiohttp/patch.py (hunk 17 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/aiohttp/patch.py:60]: Ah, stale comment, the asyncio context management handles this now

File: tests/contrib/aiohttp/test_aiohttp_client.py (hunk 19 above)
Comment (Kyle-Verhoog) [reply tests/contrib/aiohttp/test_aiohttp_client.py:45]: ah yes I forgot about HTTPretty!

Comment (brettlangdon) [reply tests/contrib/aiohttp/test_aiohttp_client.py:45]: ah, makes sense, especially since we are on such an old version for Python 2 support.

File: tests/contrib/aiohttp/test_aiohttp_client.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+
+
+@pytest.mark.asyncio
+async def test_200_request_distributed_tracing(tracer):
+    async with aiohttp.ClientSession() as session:
+        async with session.get("%s/headers" % URL) as resp:
+            assert resp.status == 200
+            headers = await resp.json()
+            for h in ("x-datadog-trace-id", "x-datadog-parent-id", "x-datadog-sampling-priority"):
+                assert h not in headers
```
Comment (brettlangdon) [inline tests/contrib/aiohttp/test_aiohttp_client.py:51]: 👍🏻 cleaner!

**Review by brettlangdon — DISMISSED** [review]
lgtm

File: tests/contrib/aiohttp/conftest.py [hunk 21]
```diff
@@ -24,7 +24,7 @@ def app_tracer(tracer, loop):
 
     @pytest.fixture
     def patched_app_tracer(app_tracer):
-        patch()
+        patch_jinja2()
```
Comment (brettlangdon) [inline tests/contrib/aiohttp/conftest.py:27]: why this change? this should still work as-is right?

we probably want to continue to verify that `patch()` does patch `aiohttp_jinja2` ?

Comment (Kyle-Verhoog) [reply tests/contrib/aiohttp/conftest.py:27]: The server instrumentation tests use an aiohttp client and so by using `patch` here we end up getting client spans in the tests for the server.

Comment (Kyle-Verhoog) [reply tests/contrib/aiohttp/conftest.py:27]: Might be worth leaving a comment here though, so I'll do that since it isn't obvious

Comment (Kyle-Verhoog) [reply tests/contrib/aiohttp/conftest.py:27]: Actually opting to not do this since I think it is pretty obvious from reading the code. I tried adding a comment but it seemed out of place.

File: tests/snapshots/tests.contrib.aiohttp.test_aiohttp_client.test_trace_query_string.json [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+    "name": "aiohttp.request",
+    "service": null,
+    "resource": "aiohttp.request",
+    "trace_id": 0,
+    "span_id": 1,
+    "parent_id": 0,
+    "type": "http",
+    "meta": {
+      "http.method": "GET",
+      "http.query.string": "k1=v1,k2%3Dv2&k3=v3&k4=v4",
```
Comment (Kyle-Verhoog) [inline tests/snapshots/tests.contrib.aiohttp.test_aiohttp_client.test_trace_query_string.json:12]: It's really strange how `=` gets encoded as `%3D` for the second param passed through the URL but that's just how the URL parsing/unparsing/stringifying seems to work 🤷

File: ddtrace/contrib/async_trace_utils.py [hunk 23]
```diff
@@ -0,0 +1,34 @@
+from ddtrace import Pin
+from ddtrace.internal.logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+def with_traced_module(func):
```
Comment (majorgreys) [inline ddtrace/contrib/async_trace_utils.py:8]: Can we move this to `trace_utils` and suffix it with `_async`? Unless you think this set of utility function for async would grow and be better off maintained separately.

File: releasenotes/notes/aiohttp-98ae9ce70dda1dbc.yaml [hunk 24]
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    aiohttp: add client integration. This integration traces requests made using the aiohttp client and includes support
+    for distributed tracing. See :ref:`the documentation<aiohttp>` for more information.
```
Comment (majorgreys) [inline releasenotes/notes/aiohttp-98ae9ce70dda1dbc.yaml:5]: After dealing with broken references in release notes, I would be inclined to avoid them here.

```suggestion
    for distributed tracing.
```

**Review by majorgreys — COMMENTED** [review]
LGTM other than these two questions

File: ddtrace/contrib/async_trace_utils.py (hunk 23 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/async_trace_utils.py:8]: We can't keep it in normal trace_utils as there would be compatibility issues with 2.7 which is why it's in its own module.

I'll add a docstring to the module to point this out though 👍

File: releasenotes/notes/aiohttp-98ae9ce70dda1dbc.yaml (hunk 24 above)
Comment (Kyle-Verhoog) [reply releasenotes/notes/aiohttp-98ae9ce70dda1dbc.yaml:5]: ah darn 😕

**Review by brettlangdon — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]
3rd ✅

**Comment by Kyle-Verhoog** [comment]
@mergifyio backport 0.x 1.0

**Comment by thehesiod** [comment]
@Kyle-Verhoog are there links for follow-up PR or issues?

**Comment by Kyle-Verhoog** [comment]
hey @thehesiod! We have not followed up yet for the `ClientResponse` or `StreamReader`. Issues for these might help us gauge the need for these features so feel free to open an issue to track those 🙂
